exwm-xsettings.el

```diff
@@ -0,0 +1,386 @@
+                 (cons (string :tag "Light Icon Theme")
+                       (string :tag "Dark Icon Theme")))
+  :initialize #'custom-initialize-default
+  :set #'exwm-xsettings--custom-set)
+
+(defvar exwm-xsettings--serial 0)
+
+(defconst xcb:xsettings:-Type:Integer 0)
+(defconst xcb:xsettings:-Type:String 1)
+(defconst xcb:xsettings:-Type:Color 2)
+
+(defclass xcb:xsettings:-Settings
+  (xcb:-struct)
+  ((byte-order :initarg :byte-order :type xcb:CARD8)
+   (pad~0 :initform 3 :type xcb:-pad)
+   (serial :initarg :serial :type xcb:CARD32)
+   (settings-len :initarg :settings-len :type xcb:CARD32)
+   (settings~ :initform
+     '(name settings type xcb:xsettings:-SETTING size
+            (xcb:-fieldref 'settings-len))
+     :type xcb:-list)
+   (settings :initarg :settings :type xcb:-ignore)))
+
+(defclass xcb:xsettings:-SETTING
+  (xcb:-struct)
+  ((type :initarg :type :type xcb:CARD8)
+   (pad~0 :initform 1 :type xcb:-pad)
+   (name-len :initarg :name-len :type xcb:CARD16)
+   (name~ :initform
+     '(name name type xcb:char size
+            (xcb:-fieldref 'name-len))
+     :type xcb:-list)
+   (name :initarg :name :type xcb:-ignore)
+   (pad~1 :initform 4 :type xcb:-pad-align)
+   (last-change-serial :initarg :last-change-serial :type xcb:CARD32)))
+
+(defclass xcb:xsettings:-SETTING_INTEGER
+  (xcb:xsettings:-SETTING)
+  ((type :initform 'xcb:xsettings:-Type:Integer)
+   (value :initarg :value :type xcb:INT32)))
+
+(defclass xcb:xsettings:-SETTING_STRING
+  (xcb:xsettings:-SETTING)
+  ((type :initform 'xcb:xsettings:-Type:String)
+   (value-len :initarg :value-len :type xcb:CARD32)
+   (value~ :initform
+     '(name value type xcb:char size
+            (xcb:-fieldref 'value-len))
+     :type xcb:-list)
+   (value :initarg :value :type xcb:-ignore)
+   (pad~2 :initform 4 :type xcb:-pad-align)))
+
+(defclass xcb:xsettings:-SETTING_COLOR
+  (xcb:xsettings:-SETTING)
+  ((type :initform 'xcb:xsettings:-Type:Color)
+   (red :initarg :red :type xcb:CARD16)
+   (green :initarg :green :type xcb:CARD16)
+   (blue :initarg :blue :type xcb:CARD16)
+   (alpha :initarg :alpha :initform #xffff :type xcb:CARD16)))
+
+(defclass xcb:xsettings:-ClientMessage
+  (xcb:icccm:--ClientMessage xcb:ClientMessage)
+  ((format :initform 32)
+   (type :initform 'xcb:Atom:MANAGER)
+   (time :initarg :time :type xcb:TIMESTAMP)      ;new slot
+   (selection :initarg :selection :type xcb:ATOM) ;new slot
+   (owner :initarg :owner :type xcb:WINDOW))      ;new slot
+  :documentation "An XSETTINGS client message.")
+
+(defun exwm-xsettings--pick-theme (theme)
+  "Pick a light or dark theme from the given THEME.
+If THEME is a string, it's returned directly.
+If THEME is a cons of (LIGHT . DARK), the appropriate theme is picked based on
+the default face's background color."
+  (pcase theme
+    ((cl-type string) theme)
+    (`(,(cl-type string) . ,(cl-type string))
+     (if (color-dark-p (color-name-to-rgb (face-background 'default)))
```

Comment:
Ok, so, this requires emacs 28.1. Should I just not support this feature on older Emacs versions?

Reply:
I wouldn't like if certain EXWM features are only available on newer Emacs versions. EXWM as a whole should work coherently on the supported Emacs version.

We could either implement the missing function for older Emacs versions or pull in Compat as dependency. Compat may be useful if we want to use other functionality. I cannot tell right now, but maybe this will be the case later (eg when we refactor some parts of the code).

Independent of this issue I suggest that we raise the dependency to 27.1 if there are no objections and support 27.1 for a few years. 27.1 is widely available on Linux distributions, including the ones with long time support.

Reply:
See https://github.com/emacs-compat/compat/blob/dff3639b70f3c9916c5397056a51cdc3e60d5668/compat-28.el#L672-L699 for the relevant code in Compat.

Reply:
My preference is 1, 2, 3 in this order. To elaborate:

1. This is the quickest solution in the short term, and easy to replace with Compat if we start using it.
2. In the long term I propose that we depend on Compat, since this will help us using newer APIs. This can be done as part of some refactoring. Compat will also be officially supported for Emacs core packages soon (see https://lists.gnu.org/archive/html/bug-gnu-emacs/2024-01/msg00660.html).
3. Removing the feature might have been my preference a while ago. But by now I use this and it really seems the easiest out of the box way to enable theme switching for the users. It would be unfortunate if users would have to reimplement this themselves.

Reply:
@Stebalien How shall we proceed? I think it is okay to merge this into the current master branch to make it available to a wider user base. The addition is low risk since it must be enabled explicitly.

Reply:
Done. Sorry, I try to limit my Emacs hacking to the weekends (try being the operative word).

Reply:
Okay, thanks! Since we cannot merge here, feel free to push this to https://github.com/emacs-exwm/exwm and close the PR here.